broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/DbSnapshotChunkReaderFactory.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.broker.clustering.atomix.storage.snapshot;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.NavigableSet;
+import java.util.TreeSet;
+
+/** Creates {@link DbSnapshotChunkReader} from a {@link DbSnapshot}. */
+final class DbSnapshotChunkReaderFactory {
+  private DbSnapshotChunkReaderFactory() {}
+
+  /** @return a static, stateless instance of the factory */
+  static DbSnapshotChunkReaderFactory factory() {
```

Comment:
I can gladly remove the singleton, but I'd rather not do work in a Ctor, esp. something that can throw an exception :scream:

Reply:
where is the difference?

Reply:
I think it's counter intuitive to have constructors throw exceptions (i.e. principle of least astonishment). I mean, exceptions can always be thrown (e.g. OOM), but here we explicitly handle/throw one :s

I guess it's a matter of personal preference though, so probably something we can decide as a team as a convention

Reply:
Don't get this reasoning normally you also check the parameters and throw IllegalArgumentException or you use Objects.requireNonNull etc. These things will throw exceptions. The reason behind this is to not create an object which are invalid.

Reply:
I'd agree if we could do RAII like in C++, but let's say we create a file directory in a Ctor, and then after throw an exception; there's no way for us to clean that file up except from outside, since we don't get a destructor :(

I'd be ok with going for a more nuanced approach, i.e. argument checking and such is totally fine in a Ctor, but doing I/O seems iffy since these can lead to exceptional...exceptions? I mean exceptions outside of our control, as opposed to getting a null value or an empty string or something.

Reply:
You could have this code in the DbSnapshot

```  @Override
  public SnapshotChunkReader newChunkReader() {
    final var directory = getDirectory();

    try {
      return new DbSnapshotChunkReader(directory, collectChunks(directory));
    } catch (final IOException e) {
      throw new UncheckedIOException(e);
    }
  }

  private static NavigableSet<CharSequence> collectChunks(final Path directory) throws IOException {
    final var set = new TreeSet<>(CharSequence::compare);
    try (var stream = Files.list(directory)) {
      stream.map(directory::relativize).map(Path::toString).forEach(set::add);
    }
    return set;
  }
```
I see no value in this singleton

Reply:
I already removed the singleton yesterday, it's just static now. Admittedly it can live in the snapshot class, I just wanted to keep it simpler. I can move it there